sympy/functions/special/bessel.py

```diff
@@ -223,11 +223,22 @@ def _eval_rewrite_as_jn(self, nu, z, **kwargs):
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         nu, z = self.args
-        arg = z.as_leading_term(x)
-        if x in arg.free_symbols:
-            return arg**nu/(2**nu*gamma(nu + 1))
-        else:
-            return self.func(nu, z.subs(x, 0))
+        try:
+            arg = z.as_leading_term(x)
+        except NotImplementedError:
+            return self
+        _, e = arg.as_coeff_exponent(x)
+
+        if e.is_positive:
+            if x in arg.free_symbols:
+                return arg**nu/(2**nu*gamma(nu + 1))
+            else:
+                return self.func(nu, z.subs(x, 0))
+        elif e.is_negative:
+            return sqrt(2)*(-(-1)**(S(3)/4)*I**nu*exp(-I*pi*nu + I*z) + \
```

Comment:
Is there a reference for this expression? Is it valid in all directions?

Reply:
Yes, It works for all `abs(z)-> oo` . I'll make sure to add references for all asymptotic approximations used in the OP.

Reply:
Can you give a reference for this expansion?

Reply:
I've added references for all bessel functions in the pr template (https://github.com/sympy/sympy/pull/23844#issue-1319631793) . So you can refer that if you want or could go through this [link](https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/02/02/02/0005/).

Reply:
There are two expressions in Wolfram. It looks like one is for the lower and the other one for the upper half-plane. I'm not sure either one of those will work for `x -> oo`. DLMF and Wikipedia would be better references.

Reply:
Yeah I've seen that wolfram sort of give 2 to 3 expressions for such cases and the first one in all cases is what we want . I had tested out limits of some expressions involving `besselj` to see if everything was in place and I didn't see anything go wrong as of now but I'll surely cross check with wikipedia and DLMF

Reply:
It's also good to have references and definitions etc in the code and docs rather than in a PR.

Reply:
That's correct , specially in case of bessel functions where almost everything related to series expansions and limits has been implemented through some reference from wolfram etc.

Reply:
> DLMF and Wikipedia would be better references.

Even Wolfram returns more than one expression based on `arg(z)` , just like what wolfram does.
<img width="297" alt="image" src="https://user-images.githubusercontent.com/87052487/184471729-a7973d79-a05b-4439-8fec-04e11fd1145c.png">

 Actually after reading through the text, I realize that the answers/outputs won't really change even if we change expressions. It's just that based on `arg(z)` we can improve the approximations furthur. For eg , let's check the limit we're trying to fix through both expressions we have.

<img width="373" alt="image" src="https://user-images.githubusercontent.com/87052487/184471818-b128e015-d5bb-4524-bc1e-ea662b40ec55.png">

We get correct answers from both expressions.
```
>>> limit(besseli(1, 3 *x)/(x *besseli(1, x)**3), x , oo)
2*sqrt(3)*pi/3
>>> limit(besseli(1, 3 *x)/(x *besseli(1, x)**3), x , -oo)
-2*sqrt(3)*pi/3
```
So after going through some resources I realize that for the output both expressions would do but in different cases 1 expression is a better approximation than the other one

Reply:
So after implementing (https://github.com/sympy/sympy/pull/23844#issuecomment-1213730451)  I guess we could try implementing two expressions for the same based on cdir (we would either come across cdir as `1` or `-1`) .

Reply:
The failing tests are unrelated I think . These tests seem to fail on every pull request as of now .

Reply:
> So after implementing ([#23844 (comment)](https://github.com/sympy/sympy/pull/23844#issuecomment-1213730451)) I guess we could try implementing two expressions for the same based on cdir (we would either come across cdir as `1` or `-1`) .

I was wrong here, as once we've implemented the comment below , we would be transforming all infinities to `S.Infinity` and dealing with the limit as x tends to `oo` . 
Hence now something like `limit(besselj(1, x), x, I*oo)` would enter the `besselj.as_leading_term()` block with a `cdir = 1`(as `I*oo` was transformed to `oo` with `cdir = -1` and `oo` will then be transformed to `0` with `cdir = -(-1)` ) .

Hence we won't be needing 2 or 3 expressions . We would just need the expression which satisfies `arg(z) = 0` which represents `x` tending to `S.Infinity` . And that's what is already implemented , hence I am quite certain we can stick to these approximations now . We can also check for limits of all bessel functions at all kinds of infinties
```
>>> limit(besselj(1, x), x, oo)
0
>>> limit(besselj(1, x), x, -oo)
0
>>> limit(besselj(1, x), x, I*oo)
oo*I
>>> limit(besselj(1, x), x, -I*oo)
-oo*I
>>> limit(bessely(1, x), x, oo)
0
>>> limit(bessely(1, x), x, -oo)
0
>>> limit(bessely(1, x), x, I*oo)
-oo*I*sign(I**(3/2))
>>> limit(bessely(1, x), x, I*oo).n()
-oo
>>> limit(bessely(1, x), x, -I*oo)
-oo*I*sign(1/sqrt(-I))
>>> limit(bessely(1, x), x, -I*oo).n()
-oo
>>> limit(besseli(1, x), x, oo)
oo
>>> limit(besseli(1, x), x, -oo)
-oo
>>> limit(besseli(1, x), x, I*oo)
0
>>> limit(besseli(1, x), x, -I*oo)
0
>>> limit(besselk(1, x), x, oo)
0
>>> limit(besselk(1, x), x, -oo)
-oo*I
>>> limit(besselk(1, x), x, I*oo)
0
>>> limit(besselk(1, x), x, -I*oo)
0
```
All results have been verified from wolfram . There's just this one limit which `limit(bessely(1, x), x, I*oo)` which wolfram straightaway approximates to `-oo` but the answer we get here through `gruntz` isn't wrong . We just might need to use `.n()` to go one step further with our approximation . I guess I should be adding this as a test in `test_limits.py`

Reply:
I've now added links/references for the expansions used in the code , hence resolving this conversation .